index join paniced on partition table when using dynamic prune

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `tbl_500` (
  `col_20` tinyint(4) NOT NULL,
  `col_21` varchar(399) CHARACTER SET utf8 COLLATE utf8_unicode_ci DEFAULT NULL,
  `col_22` json DEFAULT NULL,
  `col_23` blob DEFAULT NULL,
  `col_24` mediumint(9) NOT NULL,
  `col_25` float NOT NULL DEFAULT '7306.384497585912',
  `col_26` binary(196) NOT NULL,
  `col_27` timestamp DEFAULT '1976-12-08 00:00:00',
  `col_28` bigint(20) NOT NULL,
  `col_29` tinyint(1) NOT NULL DEFAULT '1',
  PRIMARY KEY (`col_29`,`col_20`) /*T![clustered_index] NONCLUSTERED */,
  KEY `idx_7` (`col_28`,`col_20`,`col_26`,`col_27`,`col_21`,`col_24`),
  KEY `idx_8` (`col_25`,`col_29`,`col_24`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

CREATE TABLE `tbl_600` (
  `col_60` int(11) NOT NULL DEFAULT '-776833487',
  `col_61` tinyint(1) NOT NULL DEFAULT '1',
  `col_62` tinyint(4) NOT NULL DEFAULT '-125',
  PRIMARY KEY (`col_62`,`col_60`,`col_61`) /*T![clustered_index] NONCLUSTERED */,
  KEY `idx_19` (`col_60`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci
PARTITION BY HASH( `col_60` )
PARTITIONS 1
```

```
insert into tbl_500 select -34, 'lrfGPPPUuZjtT', '{"obj1": {"sub_obj0": 100}}', 0x6C47636D, 1325624, 7306.3843, 'abc', '1976-12-08', 4757891479624162031, 0;
```

```
set tidb_partition_prune_mode='dynamic';
select tbl_5.* from tbl_500 tbl_5 where col_24 in ( select col_62 from tbl_600 where tbl_5.col_26 < 'hSvHLdQeGBNIyOFXStV' );
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

```
select tbl_5.* from tbl_500 tbl_5 where col_24 in ( select col_62 from tbl_600 where tbl_5.col_26 < 'hSvHLdQeGBNIyOFXStV' );
```

```
[2021/06/09 12:43:35.551 +00:00] [ERROR] [index_lookup_join.go:453] ["innerWorker panicked"] [conn=3855] [stack="goroutine 289897 [running]:\ngithub.com/pingcap/tidb/executor.(*innerWorker).run.func1(0x3be11f0, 0xc0014fbfc0, 0xc0017b5f30, 0xc001cd3810)\n\t/home/jenkins/ag
ent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:451 +0xd7\npanic(0x35238e0, 0xc0016f5d58)\n\t/usr/local/go/src/runtime/panic.go:965 +0x1b9\ngithub.com/pingcap/tidb/executor.prunePartitionForInnerExecutor(0x3c20ad0,
 0xc001b80000, 0x3c122a8, 0xc001b72420, 0xc0029f6640, 0xc001060e88, 0xc0017940f0, 0x0, 0x1, 0x0, ...)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:2844 +0xebe\ngithub.com/pingcap/tidb/executor.(*dataRea
derBuilder).buildIndexReaderForIndexJoin(0xc00101b5a0, 0x3be11f0, 0xc0014fbfc0, 0xc001060dc0, 0xc0017940f0, 0x0, 0x1, 0xc000c56730, 0x1, 0x1, ...)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:3583 +0x30
5\ngithub.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoinInternal(0xc00101b5a0, 0x3be11f0, 0xc0014fbfc0, 0x3c0c018, 0xc001060dc0, 0xc0017940f0, 0x0, 0x1, 0xc000c56730, 0x1, ...)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/sr
c/github.com/pingcap/tidb/executor/builder.go:3301 +0xb25\ngithub.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoin(...)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:3292\ngithub.
com/pingcap/tidb/executor.(*innerWorker).fetchInnerResults(0xc0018f5600, 0x3be11f0, 0xc0014fbfc0, 0xc001a5c100, 0xc0017940f0, 0x0, 0x1, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:6
46 +0x145\ngithub.com/pingcap/tidb/executor.(*innerWorker).handleTask(0xc0018f5600, 0x3be11f0, 0xc0014fbfc0, 0xc001a5c100, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:493 +0x155\ngi
thub.com/pingcap/tidb/executor.(*innerWorker).run(0xc0018f5600, 0x3be11f0, 0xc0014fbfc0, 0xc001cd3810)\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:470 +0x145\ncreated by github.com/pingcap/ti
db/executor.(*IndexLookUpJoin).startWorkers\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:183 +0x205\n"]
[2021/06/09 12:43:35.552 +00:00] [INFO] [conn.go:877] ["command dispatched failed"] [conn=3855] [connInfo="id:3855, addr:127.0.0.1:36550 status:10, collation:utf8mb4_0900_ai_ci, user:root"] [command=Query] [status="inTxn:0, autocommit:1"] [sql="select tbl_5.* from tbl_500
 tbl_5 where col_24 in ( select col_62 from tbl_600 where tbl_5.col_26 < 'hSvHLdQeGBNIyOFXStV' )"] [txn_mode=PESSIMISTIC] [err="runtime error: index out of range [0] with length 0\ngithub.com/pingcap/tidb/executor.(*innerWorker).run.func1\n\t/home/jenkins/agent/workspace/
build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:455\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:965\nruntime.goPanicIndex\n\t/usr/local/go/src/runtime/panic.go:88\ngithub.com/pingcap/tidb/executor.prunePartitionForInn
erExecutor\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:2844\ngithub.com/pingcap/tidb/executor.(*dataReaderBuilder).buildIndexReaderForIndexJoin\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_
master/go/src/github.com/pingcap/tidb/executor/builder.go:3583\ngithub.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoinInternal\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:3301\
ngithub.com/pingcap/tidb/executor.(*dataReaderBuilder).buildExecutorForIndexJoin\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/builder.go:3292\ngithub.com/pingcap/tidb/executor.(*innerWorker).fetchInnerResults\n\t/
home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:646\ngithub.com/pingcap/tidb/executor.(*innerWorker).handleTask\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingc
ap/tidb/executor/index_lookup_join.go:493\ngithub.com/pingcap/tidb/executor.(*innerWorker).run\n\t/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/executor/index_lookup_join.go:470\nruntime.goexit\n\t/usr/local/go/src/runtime/asm
_amd64.s:1371"]
```

### 4. What is your TiDB version? (Required)

```
Release Version: v5.2.0-alpha-16-gf81ef5579
Edition: Community
Git Commit Hash: f81ef5579551a0523d18b049eb25ab3375bcfb48
Git Branch: master
UTC Build Time: 2021-06-08 12:14:58
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

<!-- Paste the output of SELECT tidb_version() -->

--- a/executor/builder.go
+++ b/executor/builder.go
@@ -2842,6 +2842,9 @@ func prunePartitionForInnerExecutor(ctx sessionctx.Context, tbl table.Table, sch
 	}
 
 	// recalculate key column offsets
+	if len(lookUpContent) == 0 {
+		return nil, false, nil, nil
+	}
 	if lookUpContent[0].keyColIDs == nil {
 		return nil, false, nil,
 			dbterror.ClassOptimizer.NewStd(mysql.ErrInternal).GenWithStack("cannot get column IDs when dynamic pruning")

--- a/executor/partition_table_test.go
+++ b/executor/partition_table_test.go
@@ -2554,6 +2554,42 @@ func (s *partitionTableSuite) TestIdexMerge(c *C) {
 	}
 }
 
+func (s *partitionTableSuite) TestIssue25309(c *C) {
+	tk := testkit.NewTestKitWithInit(c, s.store)
+	tk.MustExec("create database test_issue_25309")
+	tk.MustExec("use test_issue_25309")
+	tk.MustExec("set @@tidb_partition_prune_mode = 'dynamic'")
+
+	tk.MustExec(`CREATE TABLE tbl_500 (
+      col_20 tinyint(4) NOT NULL,
+      col_21 varchar(399) CHARACTER SET utf8 COLLATE utf8_unicode_ci DEFAULT NULL,
+      col_22 json DEFAULT NULL,
+      col_23 blob DEFAULT NULL,
+      col_24 mediumint(9) NOT NULL,
+      col_25 float NOT NULL DEFAULT '7306.384497585912',
+      col_26 binary(196) NOT NULL,
+      col_27 timestamp DEFAULT '1976-12-08 00:00:00',
+      col_28 bigint(20) NOT NULL,
+      col_29 tinyint(1) NOT NULL DEFAULT '1',
+      PRIMARY KEY (col_29,col_20) /*T![clustered_index] NONCLUSTERED */,
+      KEY idx_7 (col_28,col_20,col_26,col_27,col_21,col_24),
+      KEY idx_8 (col_25,col_29,col_24)
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin`)
+
+	tk.MustExec(`CREATE TABLE tbl_600 (
+      col_60 int(11) NOT NULL DEFAULT '-776833487',
+      col_61 tinyint(1) NOT NULL DEFAULT '1',
+      col_62 tinyint(4) NOT NULL DEFAULT '-125',
+      PRIMARY KEY (col_62,col_60,col_61) /*T![clustered_index] NONCLUSTERED */,
+      KEY idx_19 (col_60)
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci
+    PARTITION BY HASH( col_60 )
+    PARTITIONS 1`)
+
+	tk.MustExec(`insert into tbl_500 select -34, 'lrfGPPPUuZjtT', '{"obj1": {"sub_obj0": 100}}', 0x6C47636D, 1325624, 7306.3843, 'abc', '1976-12-08', 4757891479624162031, 0`)
+	tk.MustQuery(`select tbl_5.* from tbl_500 tbl_5 where col_24 in ( select col_62 from tbl_600 where tbl_5.col_26 < 'hSvHLdQeGBNIyOFXStV' )`).Check(testkit.Rows())
+}
+
 func (s *globalIndexSuite) TestGlobalIndexScan(c *C) {
 	tk := testkit.NewTestKitWithInit(c, s.store)
 	tk.MustExec("drop table if exists p")

